packages/react/types/index.d.ts

```diff
@@ -46,6 +46,24 @@ export function withEmotionCache<Props, RefType = any>(
   func: (props: Props, context: EmotionCache, ref: Ref<RefType>) => ReactNode
 ): FC<Props & ClassAttributes<RefType>>
 
+type ReactInterpolationPrimitive =
```

Comment:
I think fundamentally what we mean is  _falsy_ or something that produce valid style information:

```ts
type ReactInterpolationPrimitive =
| Falsy 
| SerializedStyles
| CSSObject
```
Falsy in TS can be expressed like that (not the `""` and `false` instead of boolean):
```ts
type Falsy = false | "" | 0 | null | undefined
```

Any other value would be probably the result of a programming mistake and would not be interpreted correctly by Emotion anyway, so we might as wall catch them:
```ts
css={"my-custom-class"}
css={true}
```

whereas  the following could be the intended result of a boolean shortcircuit conditional:
```ts
css={""}
css={false}
```

Reply:
Hah, interesting that with `someString && somethingElse` TS actually computes the output as `"" | typeof somethingElse`! TIL

https://www.typescriptlang.org/play?#code/C4TwDgpgBAYghgGwM4igXigM0U6AfKAIkKgIAZSoA7AVwQUpqoBMJMBLKiZgKB9YDGCOACdoAgPZUkwLBIkAuKDJGcA5n0nTZcAQIhhg8ZKgwAKbCaXGUlAN5RJCCSKUr1UAL4BKdAD4oO09NKRkoMSR0OQkoADJYwMcJZ1dorx5dfUMbEDMI7xDtcIgkACYozHkoAH5EpxclSpjPKEb5DL0DIxxciNLvIA

Reply:
@Andarist yes this is pretty neat. From the quick test of my code base (a branch where I'm refactoring to use the `css` property), these types already help me a lot in finding mistakes I made.

In your example, for `res2`, I expected TS to be a little smarter with the ternary operator. I can't really understand what other value `foo` can be in the right branch other then `""` 🤷‍♂️..

So this case would generate a false negative, but to be fair I think this is a pattern that should be very rare when using Emotion. I can't really see what one would want to achieve with this where the shortcircuit conditional would be enough.

Reply:
> In your example, for res2, I expected TS to be a little smarter with the ternary operator. I can't really understand what other value foo can be in the right branch other then "" 🤷‍♂️..

Yep, same - both ternary and `&&` should really behave the same way. This was for illustration purposes - not that relevant for this PR really (I don't feel that it's a show-stopper or anything).